### PR TITLE
executor: support Chunk for NestedLoopApply

### PR DESCRIPTION
**Pull request by coocood**
For #5261

### CODE REVIEW

**Comment by coocood** [comment]
@zz-jason @winkyao PTAL

File: executor/join.go [hunk 1]
```diff
@@ -529,6 +558,36 @@ func (e *NestedLoopApplyExec) prepare(goCtx goctx.Context) error {
 	}
 }
 
+// prepare4Chunk reads all data from the small table and stores them in a slice.
+func (e *NestedLoopApplyExec) prepare4Chunk(goCtx goctx.Context) error {
+	err := e.SmallExec.Open(goctx.TODO())
```
Comment (zz-jason) [inline executor/join.go]: s/goctx.TODO()/goCtx/

File: executor/join.go [hunk 2]
```diff
@@ -529,6 +558,36 @@ func (e *NestedLoopApplyExec) prepare(goCtx goctx.Context) error {
...
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer terror.Call(e.SmallExec.Close)
```
Comment (zz-jason) [inline executor/join.go]: move this line up, we should guarantee `e.SmallExec.Close` is called even if error detected after calling `e.SmallExec.Open`

File: executor/join.go [hunk 3]
```diff
@@ -529,6 +558,36 @@ func (e *NestedLoopApplyExec) prepare(goCtx goctx.Context) error {
...
+	var selected []bool
```
Comment (zz-jason) [inline executor/join.go]: this slice is allocated every time a new outer row is fetched, we'd better buffer it outside this function.

File: executor/join.go [hunk 4]
```diff
@@ -529,6 +558,36 @@ func (e *NestedLoopApplyExec) prepare(goCtx goctx.Context) error {
 	}
 }
 
+// prepare4Chunk reads all data from the small table and stores them in a slice.
+func (e *NestedLoopApplyExec) prepare4Chunk(goCtx goctx.Context) error {
```
Comment (zz-jason) [inline executor/join.go]: s/prepare4Chunk/fetchAllInners/

File: executor/join.go [hunk 5]
```diff
@@ -572,3 +639,35 @@ func (e *NestedLoopApplyExec) Next(goCtx goctx.Context) (Row, error) {
...
+// NextChunk implements the Executor interface.
+func (e *NestedLoopApplyExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	for {
+		appendSize := mathutil.Min(e.resultChunk.NumRows()-e.cursor, e.maxChunkSize)
+		if appendSize > 0 {
+			chk.Append(e.resultChunk, e.cursor, e.cursor+appendSize)
+			e.cursor += appendSize
+		}
+		if chk.NumRows() == e.maxChunkSize {
```
Comment (zz-jason) [inline executor/join.go]: this check can be put inside `if appendSize > 0 {}`, we only change the size of `chk` in there.

File: executor/join.go [hunk 6]
```diff
@@ -500,16 +505,40 @@ func (e *NestedLoopApplyExec) fetchBigRow(goCtx goctx.Context) (Row, bool, error
...
+				return nil, false, errors.Trace(err)
+			}
+			if e.bigChunk.NumRows() == 0 {
+				return nil, false, nil
+			}
+			e.bigChunkCursor = 0
+		}
+		bigRow := e.bigChunk.GetRow(e.bigChunkCursor)
+		e.bigChunkCursor++
+		matched, err := expression.EvalBool(e.BigFilter, bigRow, e.ctx)
```
Comment (zz-jason) [inline executor/join.go]: why not evaluate outer filter using `expression.VectorizedFilter` ?

File: executor/join.go [hunk 7]
```diff
@@ -500,16 +505,40 @@ func (e *NestedLoopApplyExec) fetchBigRow(goCtx goctx.Context) (Row, bool, error
 	}
 }
 
-// prepare runs the first time when 'Next' is called and it reads all data from the small table and stores
-// them in a slice.
+func (e *NestedLoopApplyExec) fetchBigChunkRow(goCtx goctx.Context) (*chunk.Row, bool, error) {
```
Comment (zz-jason) [inline executor/join.go]: we can change this function name to `fetchSelectedOuterRow` and:
1. only return the selected outer row.
2. emit the unmatched outer row to `e.resultChunk`.
3. use the returned bool value to indicate whether there exists more outer rows.

**Comment by coocood** [comment]
@zz-jason PTAL

File: executor/join.go [hunk 8]
```diff
@@ -546,6 +610,11 @@ func (e *NestedLoopApplyExec) doJoin(bigRow Row, match bool) ([]Row, error) {
 	return e.resultRows, nil
 }
 
+func (e *NestedLoopApplyExec) doChunkJoin(bigRow chunk.Row, output *chunk.Chunk) error {
```
Comment (zz-jason) [inline executor/join.go]: useless code ?

File: executor/join.go [hunk 9]
```diff
@@ -452,14 +454,20 @@ type NestedLoopApplyExec struct {
...
-	prepared    bool
 	SmallFilter expression.CNFExprs
 	BigFilter   expression.CNFExprs
 	outer       bool
 
 	resultGenerator joinResultGenerator
 
 	outerSchema []*expression.CorrelatedColumn
+
+	bigChunk       *chunk.Chunk
```
Comment (zz-jason) [inline executor/join.go]: how about change all the `big` to `outer`, and all the `small` to `inner` ?

File: executor/join.go [hunk 10]
```diff
@@ -529,17 +564,46 @@ func (e *NestedLoopApplyExec) prepare(goCtx goctx.Context) error {
...
+// fetchAllInners reads all data from the inner table and stores them in a slice.
+func (e *NestedLoopApplyExec) fetchAllInners(goCtx goctx.Context) error {
+	err := e.innerExec.Open(goCtx)
+	defer terror.Call(e.innerExec.Close)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.innerChunkRows = e.innerChunkRows[:0]
+	for {
+		chk := e.innerExec.newChunk()
```
Comment (zz-jason) [inline executor/join.go]: these allocated inner chunks can be reused in the next `fetchAllInners()` function call.

Comment (coocood) [reply executor/join.go]: So I created this PR https://github.com/pingcap/tidb/pull/5476
to make List can be used in joinResultGenerator, and a List can be easily reused.

File: executor/join.go [hunk 11]
```diff
@@ -529,17 +564,46 @@ func (e *NestedLoopApplyExec) prepare(goCtx goctx.Context) error {
 	}
 }
 
-func (e *NestedLoopApplyExec) doJoin(bigRow Row, match bool) ([]Row, error) {
+// fetchAllInners reads all data from the inner table and stores them in a slice.
+func (e *NestedLoopApplyExec) fetchAllInners(goCtx goctx.Context) error {
+	err := e.innerExec.Open(goCtx)
```
Comment (winoros) [inline executor/join.go]: Why we use `e.innerExec.Open(goCtx.TODO())` in `prepare()`  but use `e.innerExec.Open(goCtx)` there? And the position of `defer` is also different.

Comment (coocood) [reply executor/join.go]: I just don't want to change the old code as it will be removed soon.

**Pull request merged**